---
title: "NetworkHub: An R/Bioconductor package to streamline protein-protein interaction network analysis"
author: 
- name: Lotta Wagner
  affiliation: 
  - &id1 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  email: wagnerlo@uni-mainz.de
- name: Federico Marini
  affiliation: 
  - *id1
  email: marinif@uni-mainz.de
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    theme: cosmo
    code_folding: show
    code_download: true
    global_numbering: true
    number_sections: true
editor_options: 
  chunk_output_type: console
link-citations: true
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('NetworkHub')`"
vignette: >
  %\VignetteIndexEntry{networkhub_string}
  %\VignetteEncoding{UTF-8}  
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

# About NetworkHub#TODO



# Setting up NetworkHub#TODO

```{r eval = FALSE}
# if(!requireNamespace("NetworkHub", quietly = TRUE))
#     install.packages("NetworkHub")
# BiocManager::install("NetworkHub")
``` 

If you want to install the development version from GitHub, you can alternatively run this command:

```{r}
#BiocManager::install("imbeimainz/NetworkHub")
```


setup

After installation you can load and attach the NetworkHub package as follows:

```{r}
#library("NetworkHub")
```


explain cache


Initialization of the cache of "NetworkHub"

```{r init}
#initialize_NetworkHub(nh_cachedir = "NetworkHub")
```

# Using NetworHub

To demonstrate the benefits of NetworkHub, we will use the Macrophage dataset to simulate a scenario in which NetworkHub is useful. 


## Load the data

Before we can start, we have to take a few more steps. This includes loading all the required packages:

```{r stringdb requiered packages for DE, eval = FALSE}

if(!requireNamespace("GeneTonic", quietly = TRUE))
    install.packages("GeneTonic")
BiocManager::install("GeneTonic")
```

Now we can load the SE experiment data object ("gse") from the Macrophage(<https://www.bioconductor.org/packages/release/data/experiment/html/macrophage.html>) dataset: 

```{r load from macrophage}

# dds object
library("macrophage") #demo dataset
library("DESeq2") #package to perform differential expression analysis
library("org.Hs.eg.db") #Annotation package for human genes with info about gene identifier, chromosome location, Entrez-Gene, GO-terms, ...  
library("AnnotationDbi") #Annotation package 

data("gse", package = "macrophage") #gse dataset in macrophage = summarized experiment object 
dds_macrophage <- DESeq2::DESeqDataSet(gse, design = ~ line + condition) #creation of the DESeq data object out of gse by taking corresponding design into account ( effect in condition by looking at one cell line)

rownames(dds_macrophage) <- substr(rownames(dds_macrophage), 1, 15) # make sure to only take the first 15 digits of the identifier (e.g "ENSG00000066322" and not "ENSG00000066322.14")
dds_macrophage <- estimateSizeFactors(dds_macrophage) # normalizes data
```

(OR 

Using GeneTonic we can retrieve a DE object from the macrophage dataset, which we can later integrate into a `SummarizedExperiment` (SE) object together with the previously created dataframe <dds_macrophage> containing rows with the ENSEMBLE gene symbols and columns containing the colData for the different samples and conditions. 

To compare the condition: IFNg vs naive you can use "res_macrophage_IFNg_vs_naive" DE object from GeneTonic containing rows with the ENSEMBL gene symbols and columns with the DE results ( baseMean, log2FoldChange, pvalue, ...).

```{r loadde from GeneTonic}

# res object
data(res_de_macrophage, package = "GeneTonic") #get/take a differential expressed datafile
res_de <- res_macrophage_IFNg_vs_naive #use for example an DE datafile to look into the differential expressed genes in IFNg exposed vs naive macrophages
```
)

## Differential Expression Analysis

To perform a differential expression analysis "on your own" you can run the following lines:

```{r DE}
data(gse, package = "macrophage")
dds_macrophage <- DESeqDataSet(gse, design = ~ line + condition)
# using counts and average transcript lengths from tximeta
rownames(dds_macrophage) <- substr(rownames(dds_macrophage), 1, 15)
keep <- rowSums(counts(dds_macrophage) >= 10) >= 6
dds_macrophage <- dds_macrophage[keep, ]

dds_macrophage <- DESeq(dds_macrophage)
res_macrophage_IFNg_vs_naive <- results(dds_macrophage,
                                        contrast = c("condition", "IFNg", "naive"),
                                        lfcThreshold = 1, 
                                        alpha = 0.05)
res_de <- res_macrophage_IFNg_vs_naive
de_df <- mosdef::deresult_to_df(res_de)
```

With these two objects ( dds_macrophage & res_de) we can create a SE object that contains all the dds info AND the results of a DEA and can be further used in NetworkHub :) 

```{r SE}
#SE object

dds <- dds_macrophage
de_res <- res_de
de_name <- "ifngVSnaive"

combine_dds_deres <- function(dds, de_res, de_name) {
    matched_ids <- match(rownames(res_de), rownames(dds))
    rowData(dds)[[paste0(de_name, "_log2FoldChange")]] <- NA
    rowData(dds)[[paste0(de_name, "_pvalue")]] <- NA
    rowData(dds)[[paste0(de_name, "_padj")]] <- NA
    rowData(dds)[[paste0(de_name, "_log2FoldChange")]][matched_ids] <- res_de$log2FoldChange
    rowData(dds)[[paste0(de_name, "_pvalue")]][matched_ids] <- res_de$pvalue
    rowData(dds)[[paste0(de_name, "_padj")]][matched_ids] <- res_de$padj
  return(dds)
  }
se_macrophage <- combine_dds_deres(dds = dds_macrophage, de_res = res_de, de_name = "ifng_vs_naive")

# What is this for #TODO ?
# de_res_scrambles <- de_res
# de_res_scrambles$log2FoldChange <- rnorm(17806, sd = 2)

se_macrophage

```


## PPI data retrieval 

Going from the DE genes to the proteins...

```{r get networkdata string}

# PPI df: create a dataframe containing PPIs from "Homo sapiens", version = "12.0" without annotation of the identifiers

db_stringdb_df <- get_networkdata_stringdb(species = "Homo sapiens",
                                           version = "12.0",
                                           cache = TRUE,
                                           get_annotation = FALSE,
                                           add_annotation = FALSE
                                           )



# anno df: get the annotation for "Homo sapiens"

db_stringdb_anno_df <- get_annotation_stringdb(species = "Homo sapiens",
                                               version = "12.0",
                                               cache = TRUE
                                               )

# PPI anno df: add the annotation for "Homo sapiens" from "db_stringdb_anno_df" to "db_stringdb_df" to get "db_stringdb_ppi_anno_df"

db_stringdb_ppi_anno_df <- add_annotation_stringdb(ppi_stringdb = db_stringdb_df,
                                                   anno_df = db_stringdb_anno_df,
                                                   species = "Homo sapiens")
 
```

## Creation of graph 

From there we have a SummarizedExperiment object (se_macrophage) containing the result from DE, which we want to map on a PPI network. 
Therefore the creation of a graph like object of the PPI anno df (db_stringdb_ppi_anno_df) is needed ...


```{r build_graph_stringdb}

library("igraph")

db_stringdb_igraph_object <- build_graph(graph_data = db_stringdb_ppi_anno_df,
                                         anno_df = db_stringdb_anno_df,
                                         idtype_anno = "gene_symbol",
                                         data_source = "stringdb",
                                         output_format = "igraph",
                                         min_score_threshold = "0.35",
                                         add_info_nodes = TRUE
                                         )
# returns a matrix of something around 6 x 5434 

g <- db_stringdb_igraph_object
```

Now we can fun with interactive networks...  

```{r map_SE_to_graph}

library("visNetwork")

# select an amount of genes from your de_df 
top_ones <- mapIds(org.Hs.eg.db, de_df$id[1:100], column = "SYMBOL", keytype = "ENSEMBL")
top_ones_narm <- top_ones[!is.na(top_ones)]
top_ones_in_graph <- intersect(top_ones_narm, V(g)$name)

graph_se_macrophage_string <- map_SE_on_graph(se = se_macrophage,
                                              de_name = "ifng_vs_naive",
                                              value_to_map = "log2FoldChange",
                                              id_in_graph = "attr_ensembl_id",
                                              g = db_stringdb_igraph_object
                                             )

subg <- induced_subgraph(graph_se_macrophage_string,
                         top_ones_in_graph)

visNetwork::visIgraph(subg) |>
  visNetwork::visEdges(color = list(color = "#88888888"))|>
    visNetwork::visOptions(highlightNearest = list(enabled = TRUE,
                                                   degree = 1,
                                                   hover = TRUE),
                           nodesIdSelection = TRUE)

```

Well done ! :) 

# Session Info {-}

```{r sessioninfo}
sessionInfo()
```

